### ISSUE TITLE
auto adoptation to sqlite3's configuration?

### ISSUE THREAD

**Issue by Dushistov** [rusqlite/rusqlite#747]
I am writing wrapper around C++ library that uses sqlite3.
And need to link to this C++ library statically. The problem is that I used also rusqlite.

As can I see it is possible to use `SQLITE3_STATIC`  env variable or `in_gecko` feature.
But what about sqlite3's configuration?

C++ library build sqlite3 as:
```
-DSQLITE_DEFAULT_WAL_SYNCHRONOUS=1  # Default to NORMAL sync mode on SQLite (safe with WAL mode)
-DSQLITE_LIKE_DOESNT_MATCH_BLOBS    # Optimize SQLite "like" queries
-DSQLITE_OMIT_SHARED_CACHE          # Allows SQLite to discard shared cache, eliminating some branches
-DSQLITE_OMIT_DECLTYPE              # Disable returning the declared column type from query (reduces memory usage)
-DSQLITE_OMIT_DATETIME_FUNCS        # Don't compile SQLite date functions
-DSQLITE_ENABLE_EXPLAIN_COMMENTS    # Add comment text to the result of sqlite3_explain
-DSQLITE_DISABLE_FTS3_UNICODE       # Disable FTS3 unicode61 tokenizer (not used in LiteCore)
-DSQLITE_ENABLE_MEMORY_MANAGEMENT   # Enable sqlite3_release_memory to release unused memory faster
-DSQLITE_ENABLE_STAT4               # Enable enhanced query planning
-DSQLITE_HAVE_ISNAN                 # Use system provided isnan()
-DHAVE_LOCALTIME_R                  # Use localtime_r instead of localtime
-DHAVE_USLEEP                       # Allow millisecond precision sleep
-DHAVE_UTIME                        # Use utime() instead of utimes()
-DSQLITE_OMIT_LOAD_EXTENSION        # Disable extensions (not needed for LiteCore)
-DSQLITE_ENABLE_FTS4                # Build FTS versions 3 and 4
-DSQLITE_ENABLE_FTS3_PARENTHESIS    # Allow AND and NOT support in FTS parser
-DSQLITE_ENABLE_FTS3_TOKENIZER      # Allow LiteCore to define a tokenizer
-DSQLITE_DQS=0                      # Disallow double-quoted strings (only identifiers)
```

Some options I suppose can be used also by rusqlite, like: `HAVE_UTIME` or `HAVE_LOCALTIME_R` to increase portability and performance. 

But what about other options, in bad case I always get link time errors or it is possible
to get hidden bug, when rusqlite expect some invariant from sqlite3, but this invariant is false?

I suppose the `SQLITE_THREADSAFE` is example of such options,
are any other hidden threats ?

**Comment by Dushistov** [comment]
And main question is it possible to verify build config automatically,
provide for example provide list of sqlite config options into 
rusqlite/libsqlite3-sys/build.rs and it emit build error if something wrong?

**Comment by thomcc** [comment]
We do this in both https://github.com/mozilla/application-services/ (e.g. part of mobile firefox) as well as in mozilla-central (e.g. desktop firefox) itself. 

In general rusqlite doesn't need to know about these options for the most part. What build errors are you getting?

**Comment by Dushistov** [comment]
> What build errors are you getting?
@thomcc 

My main point that I may not get any build errors.
In other words that I build and link all just fine and get random errors in runtime.
Example of such mistake may be setting `SQLITE_THREADSAFE` to wrong value and as I can see 
it doesn't cause any build time errors.

**Comment by thomcc** [comment]
Note: The 'in_gecko' feature is a bit of a hack that should be more generally expanded into a 'direct_link' feature or something.

I'd recommend in the mean time providing sqlite via rusqlite (e.g. passing in SQLITE_{LIB,INCLUDE}_DIR, and SQLITE_STATIC via environment).

--- 

I don't think rusqlite can do anything about your sqlite being potentially misconfigured prior to runtime.

**Comment by Dushistov** [comment]
> I don't think rusqlite can do anything about your sqlite being potentially misconfigured prior to runtime.

I think it can actually, as I described in the second post.
For example I set SQLITE3_CONFIG="-DSQLITE_THREADSAFE=0 -DSQLITE_..." all options that I used to build sqlite3.c amalgama and rusqlite/libsqlite3-sys/build.rs check this variable and call `panic!` if something wrong with config.

**Comment by Dushistov** [comment]
> I'd recommend in the mean time providing sqlite via rusqlite (e.g. passing in 
> SQLITE_{LIB,INCLUDE}_DIR, and SQLITE_STATIC via environment).

I will try it first of course, but I am not sure about this.
sqlite3.o is part of libBigCppLib.a , and I also link libBigCppLib.a
into my crate my_big_cpp_lib_wrapper, 
may be linking rusqlite and my_big_cpp_lib_wrapper into final executable
may cause symbol duplication error on some platform, but I hope it is not cause errors.

**Comment by thomcc** [comment]
> sqlite3.o is part of libBigCppLib.a , and I also link libBigCppLib.a

Ah, yeah, that's basically why I hacked in the `in_gecko` feature, which is for a similar use case. I'd love to do a, uh, better job there, but at the time was, well, pressed for time, and so I just did the hacky thing.

> For example I set SQLITE3_CONFIG environment variable to "-DSQLITE_THREADSAFE=0 -DSQLITE_...", in other words all options that I used to build sqlite3.c amalgama and rusqlite/libsqlite3-sys/build.rs check this variable and call panic! if something wrong with config

I now understand what you're suggesting and... Well, I worry this would just be confusing. Like, in a bundled build you'd assume this would be used to actually configure things.... And perhaps that would be valuable, but is different from what you're suggesting.

It's also a maintenance burden for a bit of an edge case. Can you think of things other than `-DSQLITE_THREADSAFE=0` for this to catch? I think you're better off just solving this via having test code that exercises stuff and thus would catch the issue.

**Comment by Dushistov** [comment]
> Can you think of things other than -DSQLITE_THREADSAFE=0 for this to catch?

R-tree flag, all compiled just fine, but then failure happens during running the application.

**Comment by thomcc** [comment]
Yeah, I don't think we want to be in the business of ensuring the SQLite build flags you used at compile time turn out to match what you end up setting up at runtime. That's a nontrivial effort if we're going to catch everything, and you can more easily catch these things via test coverage.
